### PR TITLE
Support JSON if passing application/json content type

### PR DESCRIPTION
**Pull request by sunfmin**
Mainly for testing purpose

### CODE REVIEW

**Comment by bodhi** [comment]
I thought protocol buffers had a "text serialisation" format?

**Comment by sunfmin** [comment]
Yet another Protobuf text serialisation format adding more things for developers to learn. Since everybody knows JSON, and expect to pass JSON to a http endpoint. I think support JSON is enough for testing purpose?

**Comment by bodhi** [comment]
> Yet another Protobuf text serialisation format adding more things for developers to learn.

Well, if protobuf includes a text format, then have you "learnt protobuffers" if you don' t know about it? :trollface:

But fair enough, and it doesn't look like there's a standard for it anyway.

Curious, if you're using protobuf libraries at both ends, what do you want to learn from inspecting the JSON payload?

File: prottp.go [hunk 1]
```diff
@@ -39,12 +41,16 @@ func Wrap(service Service) http.Handler {
...
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		isJson := strings.Index(r.Header.Get("Content-Type"), "application/json") >= 0
 		//func _SearchService_Search_Handler(
 		dec := func(i interface{}) error {
 			buff, err := ioutil.ReadAll(r.Body)
 			if err != nil {
 				return err
 			}
+			if isJson {
+				return json.Unmarshal(buff, i)
```
Comment (bodhi) [inline prottp.go]: Just found this: https://godoc.org/github.com/golang/protobuf/jsonpb

> This package produces a different output than the standard "encoding/json" package, which does not operate correctly on protocol buffers.

**Comment by sunfmin** [comment]
> Curious, if you're using protobuf libraries at both ends, what do you want to learn from inspecting the JSON payload?

Our developers want to use curl to poke around...

File: prottp.go (hunk 1 above)
Comment (sunfmin) [inline prottp.go]: Updated, PTAL

File: prottp.go [hunk 2]
```diff
@@ -37,10 +40,22 @@ func Wrap(service Service) http.Handler {
 	return mux
 }
 
+var marshaler = jsonpb.Marshaler{
+	EnumsAsInts:  false,
+	EmitDefaults: false,
```
Comment (bodhi) [inline prottp.go]: Suggest `true`?

Can the `.proto` file have default values in it? Ah, no: https://developers.google.com/protocol-buffers/docs/proto3#default

So if you look at the JSON output, you won't see *every* field...

File: prottp.go [hunk 3]
```diff
@@ -37,10 +40,22 @@ func Wrap(service Service) http.Handler {
...
+	Indent:       "\t",
+	OrigName:     true,
```
Comment (bodhi) [inline prottp.go]: Godoc isn't so helpful here:

```
    // Whether to use the original (.proto) name for fields.
    OrigName bool
```

Looking at the [Protocol Buffer Style Guide](https://developers.google.com/protocol-buffers/docs/style#message-and-field-names):

> Use CamelCase (with an initial capital) for message names – for example, SongServerRequest. Use underscore_separated_names for field names – for example, song_name.

So the JSON will use `underscore_separated_names` I guess.

File: prottp.go [hunk 4]
```diff
@@ -37,10 +40,22 @@ func Wrap(service Service) http.Handler {
...
+}
+
 func wrapMethod(service interface{}, m grpc.MethodDesc) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		isJson := strings.Index(r.Header.Get("Content-Type"), "application/json") >= 0
 		//func _SearchService_Search_Handler(
 		dec := func(i interface{}) error {
+			if isJson {
+				return jsonpb.Unmarshal(r.Body, i.(proto.Message))
```
Comment (bodhi) [inline prottp.go]: Suggest using [`Unmarshaler{}.Unmarshal`](https://godoc.org/github.com/golang/protobuf/jsonpb#Unmarshaler) as that will set `AllowUnknownFields` to false. It's better to be strict.

The package function is documented as: 

> This function is lenient and will decode any options permutations of the related Marshaler.

I *suspect*, but am not sure, that this means `AllowUnknownFields: true`.

File: prottp.go (hunk 3 above)
Comment (sunfmin) [inline prottp.go]: Do you think we should use `underscore_separated_names` or `camalSeparatedNames` ?

**Comment by sunfmin** [comment]
Made implementation for suggestion in: https://github.com/theplant/prottp/pull/4
Also added tests
@bodhi  PTAL

File: prottp.go (hunk 3 above)
Comment (bodhi) [inline prottp.go]: Whatever the style guide recommends.

**Pull request merged**